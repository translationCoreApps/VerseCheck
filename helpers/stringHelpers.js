/* eslint-disable no-console */
/**
 * @description Function that count occurrences of a substring in a string
 * @param {String} string - The string to search in
 * @param {String} subString - The sub string to search for
 * @returns {Integer} - the count of the occurrences
 * @see http://stackoverflow.com/questions/4009756/how-to-count-string-occurrence-in-string/7924240#7924240
 * modified to fit our use cases, return zero for '' substring, and no use case for overlapping.
 */
export const occurrencesInString = (string, subString) => {
 if (subString.length <= 0) return 0
 var n = 0, pos = 0, step = subString.length
 while (true) {
   pos = string.indexOf(subString, pos)
   if (pos === -1) break
   ++n
   pos += step
 }
 return n
}
/**
 * @description - Function that normalizes a string including whitespace
 * @param {String} string - the string to normalize
 * @preturns {String} - The returned normalized string
 */
export const normalizeString = (string) => {
  string = string.replace(/\s+/g, ' ');
  return string;
}
