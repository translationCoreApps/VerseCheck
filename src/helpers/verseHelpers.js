const ELLIPSIS = '…';

/**
 * @param {Array} verseObjects 
 * @param {Object} contextId 
 * @param {boolean} isMatch // if true, all children will be considered a match and will be included in the returned text
 */
export const getAlignedText = (verseObjects, wordsToMatch, occurrenceToMatch, isMatch=false) => {
  // TODO: FIX FOR HINDI, परमेश्वर (God), Titus 1:1, 2nd occurrence
  let text = '';
  let separator = ' ';
  let inBetween = false;
  verseObjects.forEach(verseObject => {
    if ((verseObject.type == 'milestone' || verseObject.type == 'word')) {
      if ((wordsToMatch.indexOf(verseObject.content) >= 0 && verseObject.occurrence == occurrenceToMatch) || isMatch) {
        if (inBetween) {
          separator += ELLIPSIS+' ';
          inBetween = false;
        }
        if (verseObject.text) {
          text += (text?separator:'') + verseObject.text;
          separator = ' ';
        }
        if (verseObject.children) {
          text += (text?separator:'') + getAlignedText(verseObject.children, wordsToMatch, occurrenceToMatch, true);
          separator = ' ';
        }
      } else if (verseObject.children) {
        let childWord = getAlignedText(verseObject.children, wordsToMatch, occurrenceToMatch, isMatch);
        if (childWord) {
          if (inBetween) {
            separator += ELLIPSIS+' ';
            inBetween = false;
          }
          text += (text?separator:'') + childWord;
          separator = ' ';
        } else if (text) {
          inBetween = true;
        }
      }
    } else if (verseObject.type == "text" && text) {
      if (separator == ' ') {
        separator = '';
      }
      separator += verseObject.text;
    }
  });
  return text;
};
