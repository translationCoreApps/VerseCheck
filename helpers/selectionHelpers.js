import * as _ from 'lodash';
// helpers
import * as stringHelpers from './stringHelpers';

/**
 * @description - Splice string into array of ranges, flagging what is selected
 * @param {array} ranges - array of ranges [[int,int],...]
 * @returns {array} - array of objects [obj,...]
 */
export const spliceStringOnRanges = (string, ranges) => {
  var selectionArray = []; // response
  var remainingString = string;
  // shift the range since the loop is destructive by working on the remainingString and not original string
  var rangeShift = 0
  ranges.forEach(function(range) {
    // save all the text before the selection
    var beforeSelection = remainingString.slice(0,range[0]-rangeShift);
    // save the text in the selection
    var selection = remainingString.slice(range[0]-rangeShift,range[1]+1-rangeShift);
    // save all the text after the selection
    var afterSelection = remainingString.slice(range[1]-rangeShift+1);
    if (beforeSelection) {
      selectionArray.push({text: beforeSelection, selected: false});
    }
    const stringBeforeRange = string.slice(0,range[0]);
    const occurrence = stringHelpers.occurrencesInString(stringBeforeRange, selection) + 1;
    const occurrences = stringHelpers.occurrencesInString(string, selection);
    const selectionObject = {
      text: selection,
      selected: true,
      occurrence: occurrence,
      occurrences: occurrences
    };
    selectionArray.push(selectionObject);
    // next iteration is using remaining string
    remainingString = afterSelection;
    // shift the range up to last char of substring (before+sub)
    rangeShift += beforeSelection.length;
    rangeShift += selection.length;
  })
  if (remainingString) {
    selectionArray.push({text: remainingString, selected: false});
  }
  // remove empty text from selectionArray
  return selectionArray;
}
/**
 * @description - This converts ranges to array of selection objects
 * @param {string} string - text used to get the ranges of
 * @param {array} selections - array of selections [obj,...]
 * @returns {array} - array of range objects
 */
export const selectionsToRanges = (string, selections) => {
  var ranges = [];
    selections.forEach( selection => {
      if (string !== undefined && string.includes(selection.text)) {
        const splitArray = string.split(selection.text);
        const beforeSelection = splitArray.slice(0,selection.occurrence).join(selection.text);
        const start = beforeSelection.length;
        const end = start + selection.text.length - 1;
        const range = [start,end];
        ranges.push(range);
      }
    })
  return ranges;
}

/**
 * @description - Splice string into array of substrings, flagging what is selected
 * @param {string} string - text used to get the ranges of
 * @param {array} selections - array of selections [obj,...]
 * @returns {array} - array of objects
 */
export const selectionsToSplicedString = (string, selections) => {
  var selectedStrings = [];
  const ranges = selectionsToRanges(string, selections);
  selectedStrings = spliceStringOnRanges(string, ranges);
  return selectedStrings;
}

/**
 * @description - This abstracts complex handling of ranges such as order, deduping, concatenating, overlaps
 * @param {array}  ranges - array of ranges [[int,int],...]
 * @returns {array} - array of optimized ranges [[int,int],...]
 */
export const optimizeRanges = (ranges) => {
  var optimizedRanges = []; // response
  ranges = _.sortBy(ranges, function(range) { return range[1] }); // order ranges by end char index as secondary
  ranges = _.sortBy(ranges, function(range) { return range[0] }); // order ranges by start char index as primary
  ranges = _.uniq(ranges); // remove duplicates
  // combine overlapping and contiguous ranges
  var _range = [];
  ranges.forEach( (range, index) => {
    if (range[0] >= _range[0] && range[0] <= _range[1]+1) { // the start occurs in the running range and +1 handles contiguous
      if (range[1] >= _range[0] && range[1] <= _range[1]) { // if the start occurs inside running range then let's check the end
        // if the end occurs inside the running range then it's inside it and doesn't matter
      } else { // it doesn't occur inside the running range
        _range[1] = range[1]; // extend running range
      }
    } else { // the start does not occur in the running range
      if (_range.length != 0) optimizedRanges.push(_range) // the running range is closed push it to optimizedRanges
      _range = range; // reset the running range to this one
    }
    if (ranges.length === index + 1 && _range[1] - _range[0] >= 0) { // this is the last one and it isn't blank
      optimizedRanges.push(_range); // push the last one to optimizedRanges
    }
  });
  return optimizedRanges;
}

/**
 * @description - This converts ranges to array of selection objects
 * @param {string} string - text used to get the ranges of
 * @param {array} ranges - array of ranges [[int,int],...]
 * @returns {array} - array of selection objects
 */
 export const rangesToSelections = (string, ranges) => {
   let selections = [];
   ranges.forEach( range => {
     const start = range[0], end = range[1]; // set the start and end point
     const length = end - start + 1; // get the length of the sub string
     const subString = string.substr(start, length); // get text of the sub string
     const beforeText = string.substr(0, start); // get the string prior to the range
     const beforeMatches = stringHelpers.occurrencesInString(beforeText, subString); // get occurrences prior to range
     const occurrence = beforeMatches + 1; // get number of this occurrence
     const occurrences = stringHelpers.occurrencesInString(string, subString); // get occurrences in string
     const selection = {
       text: subString,
       occurrence: occurrence,
       occurrences: occurrences
     };
     selections.push(selection);
   })
   return selections;
 }

/**
 * @description - This abstracts complex handling of selections such as order, deduping, concatenating, overlapping ranges
 * @param {string} string - the text selections are found in
 * @param {array}  selections - array of selection objects [Obj,...]
 * @returns {array} - array of selection objects
 */
export const optimizeSelections = (string, selections) => {
  let optimizedSelections; // return
  var ranges = selectionsToRanges(string, selections); // get char ranges of each selection
  ranges = optimizeRanges(ranges); // optimize the ranges
  optimizedSelections = rangesToSelections(string, ranges); // convert optimized ranges into selections
  return optimizedSelections;
}
