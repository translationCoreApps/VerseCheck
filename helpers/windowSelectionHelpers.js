import * as selectionHelpers from './selectionHelpers';
/**
 * @description - Gets the currently selected text from the Web UI
 * @param {String} verseText - the text that the selection should be in
 * @return {Object} - the selection object to be used
 * Implementation notes on why you can't just use the window.getSelection()
 * getSelection is limited by same innerText node, and does not include span siblings
 * indexOfTextSelection is broken by any other previous selection since it only knows its innerText node.
 */
// TODO: Make this into smaller functions.
export const getSelectionText = (verseText) => {
  // replace more than one contiguous space with a single one since HTML/selection only renders 1
  verseText = selectionHelpers.normalizeString(verseText);
  // TODO: Determine if a user can even get this far without being logged in.
  // if (!this.props.loginReducer.loggedInUser) {
  //   this.props.actions.selectModalTab(1, 1, true);
  //   this.props.actions.openAlertDialog("You must be logged in to make a selection");
  //   return;
  // }
  // windowSelection is an object with lots of data
  let windowSelection = window.getSelection();
  // get the current text selected
  let selectedText = windowSelection.toString();

  // do nothing since an empty space was selected
  if (selectedText !== '') {
    // get the text after the presceding selection and current span selection is in.
    let selectionRange = windowSelection.getRangeAt(0)
    // get the character index of what is selected in context of the span it is in.
    let indexOfTextSelection = selectionRange.startOffset;
    // get the container of the selection, this is a strange object, that logs as a string.
    let textContainer = selectionRange.commonAncestorContainer;
    // get the parent span that contains the textContainer.
    let textSpan = textContainer ? textContainer.parentElement : undefined;
    // get all of the text in the selection's container similar to the span's innerText.
    let textSpanContent = textSpan.innerText;
    // get the text presceding the selection but after the selection just prior to it.
    let postPrescedingText = textContainer ? textSpanContent.slice(0,indexOfTextSelection) : '';
    // start with an empty string to prepend to for text presceding current span selection is in.
    let prescedingText = '';
    // if we have a span that holds text, see what presceding text we can extract.
    if (textSpan) {
      // get the previous sibling to start the loop
      let previousSibling = textSpan.previousSibling;
      // loop through previous spans to get their text
      while (previousSibling) {
        // prepend the spans innerText to the prescedingText
        prescedingText = previousSibling.innerText + prescedingText;
        // move to the previous span, if none, it ends the loop
        previousSibling = previousSibling.previousSibling;
      }
    }

    // There can be a gap between prescedingText and current selection
    let textBeforeSelection = prescedingText + postPrescedingText + selectedText;
    // get the occurrence of the selection
    let occurrence = selectionHelpers.occurrencesInString(textBeforeSelection, selectedText);
    // get the total occurrences from the verse
    let occurrences = selectionHelpers.occurrencesInString(verseText, selectedText);
    let selection = {
      text: selectedText,
      occurrence: occurrence,
      occurrences: occurrences
    };
    // add the selection to the selections
    return selection;
  }
}
