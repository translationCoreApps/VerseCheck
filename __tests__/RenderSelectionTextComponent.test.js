/* eslint-env jest */
import React from 'react';
import RenderSelectionTextComponent from '../src/components/RenderSelectionTextComponent';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { getCurrentWindowSelection } from '../src/helpers/windowSelectionHelpers';

describe('RenderSelectionTextComponent component Tests', () => {
  let props;

  beforeEach(()=> {
    props = {
      actions: {
        changeSelectionsInLocalState: jest.fn(),
        openAlertDialog: jest.fn(),
      },
      mode: 'select',
      verseText: 'verse text',
      selections: [{text: 'text'}]
    };
  });

  test('Check RenderSelectionTextComponent component', () => {
    const wrapper = shallow(<RenderSelectionTextComponent {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.text()).toEqual(props.verseText);
  });
});
