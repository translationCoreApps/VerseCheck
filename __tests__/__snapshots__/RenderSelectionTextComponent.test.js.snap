// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderSelectionTextComponent component Tests Check RenderSelectionTextComponent component 1`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
>
  <span
    key="0"
    onClick={[Function]}
    style={
      Object {
        "color": "black",
      }
    }
  >
    verse text
  </span>
</div>
`;
